server/botany/tests/test_download.py

```diff
@@ -0,0 +1,245 @@
+import io
+import contextlib
+import pathlib
+from unittest.mock import patch
```

Comment:
I'm a bit allergic to patching unless absolutely necessary.  (I find mocking and patching to be quite brittle, and that it hides bugs because you end up testing a system that's not in a realistic state.)

You're patching `get_bots()` and `get_bots_for_api()`, but these are functions whose return value we can control by creating bots via the factory functions that you've seen.  I'll merge this PR if you choose not to change this, but I may well change it myself soon.

Reply:
I'm still trying to learn how to get the balance right with patching/mocking so I would definitely appreciate input on this.

I do want to point out that I've not patched anything in the tests for get_bots, which is the only code in download.py that interacts with the ORM. I thought it would be acceptable to isolate the other tests from the ORM because I've already attempted to confirm that get_bots returns what I expect it to. Does this seem reasonable?

Reply:
In my experience, the problem with this kind of patching is that you need to make sure that: 

1) `get_bots` is patched correctly today.
1)  `get_bots` continues to be patched correctly in future.

There's no way for us to check 1. holds without very careful reading of the code for `get_bots`.  And if we ever change what `get_bots` returns, we'll have to watch out for 2., and check for anything that patches it, and change that too.  This is significant extra work and extra risk for limited gain.

There certainly are places in Python where mocking/patching makes good sense, but you need to be clear about what problem you're using it to solve.  A couple of examples:

1) If there is an external service (like an STMP server) that your code talks to, and you want to test that your code responds to errors correctly, it absolutely makes sense to mock your interactions with the service.
1) If your test suite is slow because it's spending a lot of time hitting the database, and you have tests that (a) create objects in the database; (b) load objects from the database; (c) process the objects; (d) check that the processing did the right thing, then it may make sense to introduce patches so that you can skip (a) and (b), speeding things up.  But doing this well requires planning, otherwise you'll end up with patched objects throughout your test suite and suffer from the problems of extra work and extra risk outlined above.

Does this make sense?

As I said, I'm happy to accept this PR as is, but I will probably rewrite the tests myself as part of #20.

Reply:
Thanks @inglesp that all makes sense. I guess without testing the patches, which sounds a bit far-fetched for this simple case, I can see how it could be a problem.

Hopefully commit 4776750 is an improvement!

Reply:
Yes, I prefer 4776750c4585a487ffe5c1a77a17a42d9f0287bd.  Thanks!